### DISCUSSION TITLE
March 17, 2021 working meeting

### DISCUSSION THREAD

**Discussion by tonyfast** [jupyter/accessibility#31] [Meeting]
Welcome to our second hands on accessibility community meeting. These events are for learning and sharing knowledge about accessibility within our diverse, growing community. On alternate week,s @isabela-pf will run our project meetings.

> if you are seeing this issue, please react to issue with each thing you want to participate in, you should choose more than 1.

- :heart: i'm happy to be here!
- :eyes: i want learn about accessibility by reviewing existing issues and pr's
  - review and discuss https://github.com/jupyter/accessibility/issues/27
- :smile: i want to work on the [readme](readme.md) with @isabela-pf _only requiring markdown knowledge_
  - add badges to gitter, google calendar, discourse, other meetings folks can go to, edit the contents of the readme
- :tada: i want help close last meetings issue by accumulating the links and references into the Documentation discussion.
  - https://github.com/jupyter/accessibility/issues/23 -> https://github.com/jupyter/accessibility/discussions/29

we'll split into groups and ~~work/learn~~**have fun** together.

please use this issue to add meeting notes by adding comments.

**Comment by saulshanabrook** [comment]
# Notes

Tony's ideas:

* Good to work on the README?
* Take meeting notes from last week, add to discussions?

---

Saul: How do we know what to focus on? 

Isabella: Blindness is to capture a lot of things. We currently have this project board: https://github.com/jupyter/accessibility/discussions/32 Our goal is WCAG compliance. We cant test some of them, because you cant get testing. 

Nick: People use nonfree screen readers. Screen readers are expensive, if you are deaf and blind and you are using a braille screen reader, that machine is thousands of dollars. Can't imagine testing against it.

Isabela: We have had the privilege that other people have tested already. Three-year-old issues, that we haven't given attention too. 

Tony: JAWS. NVDA, and WAVE are testing schools. These test against standards, not screen readers. 

Nick: Yeah they are not representing a screen reader, but aren't pretending to. A thing that just came up is icons. All the icons in lab, don't have any semantic meaning to them. Just have photos, not much metadata. Reader will say `UI-componentns:refresh`. People crank this up to 200 wpm, so they go very quickly.

---

Tony: Resource throughout all of this is [a11y project](https://www.a11yproject.com/). One of their suggestions is to have a no mouse day! 

---

Nick: Long running issue is losing output, because of closing long running job. If we had this on the server, would make a lot of things possible. If you don't have to talk to a websocket. Would be nice to deliver a first class jupyter experience, with widgets, to the terminal, because its more robust here. If ipython could edit a notebook, that would edit and run, and see output of, would allow us to describe some of these things. 

Tony: The terminal has been accessible for a while, including tools like hypterm. 

---

Tony: Want to split into two groups. Now that binder is merged, easy to do README things. We can help fill out the README, have PR open for this.

Other one, is that we have a lot of good notes in our issue from last one. Would be helpful to move this to a discussion. We have been getting a lot of good documentation resources, in short term, would be to put into a discussion. 

The two groups will split off, one group to do README, the other triage?

---

Isabela: This is the learning meeting, just FYI. Other meeting is for project management. They switch every week.

Tony: Actually MJ and I will stay here and look at issue!


## Other group notes

Notes from subgroup:


Issues: https://github.com/jupyterlab/jupyterlab/pull/8832

Color contrast tooling: https://www.tpgi.com/color-contrast-checker/
<img width="1069" alt="Screen Shot 2021-03-17 at 2 03 48 PM" src="https://user-images.githubusercontent.com/1186124/111515938-c0069d80-8729-11eb-9e93-d8fd490387f3.png">


Current at command palette proposals. For example, when we have darker highlight over text boxes, we have some issues:

<img width="473" alt="Screen Shot 2021-03-17 at 2 08 10 PM" src="https://user-images.githubusercontent.com/1186124/111516381-35726e00-872a-11eb-8872-3d1d69780053.png">

So I was proposing it has an outline. All you would have to do is which variable it is. Why is it grey when others are blue?

I have proposed solutions for the command palette, feel free to review it. If you want to look at putting things into code, I can help there. Easiest place would be to start there.

Other option would be to review other parts of the UI. The command palette has the most violations. You could go and audit some other region! Or if you wanna jump straight into code. If you want practice doing a PR, you could get practice, just changing a part of it, like change outline.

### README

Tony: We need some badges, we need some color. They are going to have a bunch of badges, for gitter, the calendar, this is a matter of learning some markdown, to make some pictures and links. If you all could aggregate find all the resources, so they know when the meetings are, they know where the chat is.

There is a great resource for making a README: https://github.com/noffle/art-of-readme

Badges: Here's an example of how those badges get applied in a tiny markdown readme: 
https://github.com/isabela-pf/jupyter-communitycalls

As new people, what is that you need to know? Comment on the PR, things I wish I saw when I showed up at the repo. we have lost all that, since we have been here a while. 

Isabella: Would be helpful to have some direction, for anyone to comment.

Sara: What is the difference between a badge and a hyperlink in the text?

Isabella: Functionally the same, just break it up visually. 

Quinten: I didn't know how much freedom we really have, but it doesn't seem like that beginner's eye can be helpful.

Sara: We should put badges in our readme too!

Tony: Demoing how to find markdown Copy Gitter badge from JupyterLab readme source. Copy and paste into issues. Remember alt text, after link! Reference: https://daringfireball.net/projects/markdown/syntax#img

**Reply by tonyfast** [reply]
these notes are killer. thanks for doing this saul.

**Comment by isabela-pf** [comment]
In response to #27 How/why are we prioritizing what we are for JupyterLab?
- WCAG standards are the end goal, but how do we get there?
- Screen reader and needs of blind users usually require the most work on our part, so accessibility standards often advocate for starting with this work and moving from there. We are following that recommendation.
- People have tested things for us. Some people have even proposed solutions. We are prioritizing that as well. jupyterlab/jupyterlab #9399 is an often linked example of that. So is the [accessibility project board](https://github.com/orgs/jupyterlab/projects/1).
- JupyterLab isn't even fully testable as is because it is missing some foundational parts (see jupyterlab/jupyterlab #9491). We also need to prioritize this because it opens doors for us to test and meet WCAG.

**Reply by manfromjupyter** [reply]
Disagree with starting with blind users. Blind users can't be supported or even tested until ambulatory are working to start and if it only takes a week to support `1/7th` of the disability groups by tackling another category, why save it for last? We need movement. If we stop a month or three in from now and don't finish or we lose our ~2 typescript resources, what do we have to show for it? We would have made it better but it's still just as inaccessible to all groups it was before.

I admit, I should have created different subtickets and/or in a different order, I just had different expectations in terms of speed I have since readjusted. Will see if I might be able to get authorized to take some time to code towards this.

I think we need to get the color blind users all squared away that's a very basic 1 ticket worth of work to my knowledge (detailed below from [#9399](https://github.com/jupyterlab/jupyterlab/issues/9399)). The person that fixes it just needs to know CSS, just use the [Tota11y bookmarklet](https://khan.github.io/tota11y/) for it to output exactly what to change each of those too. Then get the Low Vision users done at the same time as supporting mobile-tablet support. I started working on detailing the fixes for this last Wednesday but never finished making it. It's really not much, almost entirely CSS. Then finish ambulatory by doing 3 things: add tabindexing, make the content index in a logical order (which will just be moving blocks of code or hardcoding the indexes), and then have the focus always visually exposed (to whatever you're tabbed to, just add CSS). All of that together lays the groundwork and finishes a bit of Blind users as well. We already have users with Deaf/Hard of Hearing, Seizure, and Cognitive disabilities squared away, so then we can tackle Blind users.

I advocated for the completion of the landmarks ticket [#9411](https://github.com/jupyterlab/jupyterlab/issues/9491) and [#9622](https://github.com/jupyterlab/jupyterlab/pull/9622) (purely for blind users) because it completely knocked off a section we had to do, was the least we could do to help blind users for now, and allowed us to then splinter the rest of work for blind users from there because we could navigate around now. Hard truth, blind user support will require refactoring every panel and modal of the product and making sure we are revealing what needs to be revealed, hide what is decorative, and make sure it's efficiently done. We can get most of the hard work out of the way by supporting ambulatory users but it is my opinion that we need to finish the other disability groups first because they'll be so quick. I think there is plenty of work to go around and we can have people work on multiple sections including blind solutions but think our voice and/or direction should be that we're trying to support as many people as fast as we possibly can.

## Work that is left (and how to do it)
### Color Blindness
**For color blindness,** you do these 8 CSS changes as detailed in  [#9399](https://github.com/jupyterlab/jupyterlab/issues/9399), and we can say we support 8% of the male population and 0.5% of the female population that we didn't previously (Source: [colorblindawareness.org](https://www.colourblindawareness.org/colour-blindness/#:~:text=Colour%20Blindness,-What%20is%20colour&text=Colour%20(color)%20blindness%20(colour,200%20women%20in%20the%20world.))).
> #### Color, Contrast, and Zoom
> Issue Area # 1
> * **Expectation**: The default text and background size and colors provide sufficient contrast. (WCAG Criteria 1.4.3 (AA)).
> * **Observed for Default theme/OOTB Functionality**:
>   
>   * File Browser/settings highlighted field have a 3.12 contrast ratio (white on light blue background). Needs to be contrast ratio of 4.5.
>   * Kernal Sessions “Shut Down” buttons have a 2.16 contrast ratio (bright orange on white) out of 4.5.
>   * Search section disabled console actions have a 2.68 contrast ratio out of 4.5 (gray on white background).
>   * Notebook “[ ]:” text is 4.26 contrast ratio (light blue/green on white). Needs to be 4.5.
>   * Syntax highlighting fails in numerous places. The # 258f8f is 3.08 on red error background, the # 627d12 a 2.85, the # a1a6b2 a 1.93 – make instead a # 007373, # 895f00, and # 5d6776 respectively.
>   * Settings tabs “1 {“ text and file endings featuring green on white/gray are 2.59/2.85. Make it #008a00 instead.
>   * Syntax highlighter color # 8888ff is 2.99, use # 6464f1.
> 
> Issue Area # 2
> * **Expectation**: Interface components and graphical objects that convey information have sufficient contrast with the background and surrounding objects (WCAG Criteria 1.4.11 Not-Text Contrast (AA)).
> * **Observed**: File search section search icon is 1.87 contrast (dark gray magnified glass on blue background) and needs to be a 3.0 or higher.

### Low Vision
- Entirely CSS (pretty much)
Been swamped as of late but will make a ticket by next Wednesday meeting. Took the screenshots, just need to add arrows and notes to them. In summary, just make it so all content is accessible and only a vertical scrollbar appears for everything except the code blocks. Each of the panels have predefined heights set on their inner elements. Change to `height: 100%` or auto or 100% after you get to a browser width of 800px or something if the Jlab veterans are worried about losing the functional it might currently have. Then get rid of a few `overflow: hidden`'s so we can scroll down on the panels and notebook section. Super easy. Only hard part I see which just requires some creativity and guts is only one panel can be open at any time after a certain size and minimizes the other but visually indicates the tab that was close and how to toggle back OR the side bar panels (left and right side) are overlaid over the notebook editor/center panel when opened. 

### Ambulatory
3 things: 
- add tabindexing (typescript HTML changes or just attribute additions)
- visually expose focus at all times (pure CSS)
- make the content index in a logical order (will require debate among Jupyter leads and will involve most likely moving of elements around (as they appear in the DOM). Maybe also some JS and CSS to fix. No rewriting of code, just tedious.

### Blind
Refactor every panel and modal of the product to ensure they have the necessary aria elements, speak the correct information to the user, in an as efficient way as possible, and with none of the noise. (Requires HTML, JS, and TypeScript knowledge and have to touch just about every section with text or icons in the code - so will take a while even just to merge). 

Comes out to like 30 things
- Aria Labels to all elements that need them
    - Then get those translated if different than something already exposed
- Properly indicate what's a form with form role if we can't switch to proper HTML `<form></form>
- Add landmark for search in File Browser `role="search"`
- aria-hidden="true" to ever icon that provides no information.
- make all error messages throw a` role="alert"` to interrupt the user to inform them of errors they've run into.
- Add `alt=" "` to all images
- Add description and title tags to the few SVGs
- Audit all links to say if they're internal or external links. In addition, if external, let them know about where they are going or what they might see (needs aria-labels)
- Make sure all inputs provide hints for the screenreader and are properly formatted to limit their character to certain types of entries or types of input if applicable
- Audit error messages to ensure they provide exact reasoning for what's wrong and how to fix the issue.
- Make it so all buttons open up with both Space AND Enter key
- aria-expanded, aria-popup, alerts during and after certain actions have finished loading, caption tags to the few tables, restructure headings to a proper order no duplicate H1's, need at least 1 H1, no H3's under H1's, and more.

**Comment by manfromjupyter** [comment]
Just because I mentioned it earlier, wanted to bring up the creation of the 2 tickets to address Colorblindness and Low Vision support. One bite sized ticket to completely (or 99%) knock out the WCAG support for these disabilities so we can just focus on Ambulatory and Blindness.
- Full Color Blind Support ticket: jupyterlab/jupyterlab#10008
- Full Low Vision (Mobile/Tablet Support) ticket: jupyterlab/jupyterlab#10004
